### PR TITLE
Add Index to IO Batch

### PR DESCRIPTION
**Pull request by junranhe**
在mxioiter添加了每个batch中添加了index，用于在python中查找其他辅助信息

### CODE REVIEW

File: src/io/iter_prefetcher.h [hunk 1]
```diff
@@ -80,6 +80,9 @@ class PrefetcherIter : public IIterator<DataBatch> {
                         batch.data[i].FlatTo2D<cpu, real_t>());
           (*dptr)->num_batch_padd = batch.num_batch_padd;
         }
+        for (size_t i = 0; i < batch.batch_size; ++i) {
+          (*dptr)->index[i] = batch.inst_index[i];
```
Comment (tqchen) [inline src/io/iter_prefetcher.h]: use std::copy

File: src/c_api/c_api.cc [hunk 2]
```diff
@@ -838,6 +838,22 @@ int MXDataIterGetLabel(DataIterHandle handle, NDArrayHandle *out) {
   API_END();
 }
 
+int MXDataIterGetIndex(DataIterHandle handle, uint64_t *out_index) {
```
Comment (tqchen) [inline src/c_api/c_api.cc]: Try to return in one function, 
```
int MXDataIterGetIndex(DataIterHandle handle, uint64_t **out_index, size_t *out_size)
```

**Comment by tqchen** [comment]
Thanks for the contribution! This is helpful. Given that there is another IO refactor going on #153  I will not merge it for now. 
We can rebase from the on going refactor repo and push changes to there, or wait until #153 is merged in

@pluskid

**Comment by tqchen** [comment]
Sorry It is  #468 I referred to

**Comment by pluskid** [comment]
Sorry for holding this back... I will implement a simple peak method hopefully today or tomorrow and get that pr merged.

Meanwhile: could you explain a little bit more on The use case of this interface? I'm thinking maybe similar information could be simply obtained by chaining the data iter with the python enumerate function. Also other data iter needs to be modified to add the same interface.

**Comment by tqchen** [comment]
I think what was wanted is because each image have a unique index, so we can use it to remap labels. Say we store the imid -> bounding box in an independent map on python, and want to attach that information to the iter, while still being able to use the preproc pipeline

**Comment by pluskid** [comment]
I see. Thanks!

**Comment by tqchen** [comment]
@junranhe  #468 is merged, can you rebase and update this PR? Thanks

**Comment by junranhe** [comment]
不好意思，昨天周日没上git，我会尽快重新提交的

**Comment by junranhe** [comment]
@tqchen
int MXDataIterGetIndex(DataIterHandle handle, uint64_t **out_index, size_t *out_size)返回后怎么用 uint64_t **out_index 直接转换为numpy.array?我没有找到array直接用指针初始化的方法哦，所以之前才分为MxDataIterGetBatchsize() 和 MXDataIterGetIndex(),这样能先初始化bumpy.array，再把数据拷到里面

**Comment by tqchen** [comment]
hmm, I think you are right, that is easier.

https://github.com/dmlc/mxnet/blob/master/python/mxnet/base.py#L131 This function should also be modified to do what was needed. But do need to copy here as well instead of sharing.

**Comment by junranhe** [comment]
@tqchen 在MXDataGetIndex里面 DataBatch为const的，所以返回*out_index = db.index.data()是需要const_cast，介意不？我一直觉得强制转换不优雅

**Comment by junranhe** [comment]
@tqchen  请问r_test 通不过，是要注意些什么吗？R的环境细节我不熟悉。。。

**Comment by thirdwing** [comment]
@junranhe Because you delete the `rpkg` in Makefile, it is used to build the R package.

**Comment by junranhe** [comment]
@thirdwing 好，改了，同时想问下可不可以吧 MXNET_CUDNN_PATH这个参数放回来？实际开发环境有多套cudnn，指定路径比较方便:)

**Comment by junranhe** [comment]
@tqchen  这个pr还有需要吗？要不撤了，有需要我在重新pr，rebase得比较danteng，还有checkfail 在 一个cpp_test里面，感觉跟io没什么关系啊？我比较忙没时间深究。。。

**Comment by tqchen** [comment]
Sorry about the delayed reply, this is merged

**Pull request merged**